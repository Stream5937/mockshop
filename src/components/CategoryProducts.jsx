import { useOutletContext, Link, useParams } from "react-router-dom";
import style from "../styles/category.module.css";

export default function CategoryProducts() {
  const context = useOutletContext();
  const { products, error, loading } = context[0];
  const [cart, setCart] = context[1];
  const [removeFromCart, clearCart, sumCart] = context[2];
  const [quantity, setQuantity] = context[3];

  const { categoryName } = useParams();

  const filtered = products.filter((p) => p.category === categoryName);

  if (filtered.length === 0) return <p>No products found in this category.</p>;

  return (
    <div>
      <h2>{categoryName}</h2>
      <div className={style.gridContainer}>
        {filtered.map((product) => (
          <div id={style.product} key={product.id}>
            <img
              id={style.imgCategory}
              src={product.image}
              alt={product.title}
            />
            <h3>{product.title}</h3>
            <p>
              <b>${product.price}</b>
            </p>
            <p>{product.description.substring(0, 80)}...</p>
            <p>
              ‚≠ê {product.rating.rate} ({product.rating.count} reviews)
            </p>
            <Link to={`/product/${product.id}`}>View Product</Link>
          </div>
        ))}
      </div>
    </div>
  );
}
