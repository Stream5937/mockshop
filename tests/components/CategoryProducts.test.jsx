import { describe, it, expect, vi, beforeEach } from "vitest";
import { renderWithRouter } from "../utils/renderWithRouter.jsx";
import CategoryProducts from "../../src/components/CategoryProducts.jsx";

// Mock react-router-dom
vi.mock("react-router-dom", async () => {
  const mod = await vi.importActual("react-router-dom");

  const mockProduct = {
    id: 1,
    title: "Test Product",
    category: "Test Category",
    price: 10,
    description: "This is a test product description",
    rating: { rate: 4.5, count: 20 },
    image: "test.jpg",
  };

  const mockContext = [
    { products: [mockProduct], error: null, loading: false }, // context[0]
    [[{ item: mockProduct, number: 1 }], vi.fn()], // context[1]
    [vi.fn(), vi.fn(), vi.fn()], // context[2]
    [1, vi.fn()], // context[3]
  ];

  return {
    ...mod,
    useOutletContext: () => mockContext, // return array
    useParams: () => ({ categoryName: "Test Category" }), // mock params
  };
});

describe("CategoryProducts component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders with products in the category", () => {
    const { asFragment, getByText } = renderWithRouter(<CategoryProducts />);

    // Check snapshot
    expect(asFragment()).toMatchSnapshot();

    // Check important UI bits
    expect(getByText("Test Category")).toBeInTheDocument();
    expect(getByText("Test Product")).toBeInTheDocument();
    expect(getByText(/\$10/)).toBeInTheDocument();
    expect(getByText(/‚≠ê 4.5/)).toBeInTheDocument();
  });
});
