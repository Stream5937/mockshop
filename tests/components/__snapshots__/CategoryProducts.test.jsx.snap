// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryProducts component > renders with products in the category 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Test Category
    </h2>
    <div
      class="_gridContainer_41d463"
    >
      <div
        id="_product_41d463"
      >
        <img
          alt="Test Product"
          id="_imgCategory_41d463"
          src="test.jpg"
        />
        <h3>
          Test Product
        </h3>
        <p>
          <b>
            $10
          </b>
        </p>
        <p>
          This is a test product description...
        </p>
        <p>
          ‚≠ê 4.5 (20 reviews)
        </p>
        <a
          data-discover="true"
          href="/product/1"
        >
          View Product
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
